### PR TITLE
Feature: Add P2SH support - Core Protocol Development Proposal 001 (PR #873)

### PR DESCRIPTION
**Pull request by friendofjeroz**
**This development is for the Core Protocol Development Proposal 001** 
Found at - https://gateway.ravencoinipfs.com/ipfs/QmawFRdE8WzgshLK9fGpqMaHu2HUKEvSJcWPqVd9iiU7zs

The following has been added to the source code:
- Support for P2SH with the asset type transactions
- Unit tests for the new P2SH and asset integration
- Functional tests for the new P2SH and asset integration (Using console command, and also createrawtransaction paths)
- P2SH Assets has been added to a BIP16 forking mechanism that starts voting on the following days. Testnet = UTC: Mon Mar 01 2021 19:00:00, Mainnet = UTC: Thu Apr 01 2021 18:00:00
- Add updated checkpoints for Mainnet and Testnet to increase sync speed
- Removed lots of duplicate code around different asset types where the same code was being used in switch-case statements
- Reorganized code and moved some code into functions to remove duplicate code

If any changes are needed, please place comments on this PR, and I will get to them when I have some spare time. 

Please verify with Jeroz before sending payment. Would like to confirm address before it is sent. (Always got to double check that shiz)

Payment can be sent to: **RUGgLoq7eosejqsYuVkPgs6A1CNFF1DdWh**
I'd like 10% of this payment to go to Jeroz as a finders fee - Thanks Jeroz! P.S - he didn't know about this. 

If you love the work I did, you can send extra donations to **RKGUfpTntLGRRmwvuWE1vTnskfEG1DKc2h**

Thanks Hope You Enjoy

### CODE REVIEW

**Comment by friendofjeroz** [comment]
Looks like the pipeline is having some issues. This shouldn't be related to the submitted code. Have a good day!

**Comment by TronBlack** [comment]
It's not related to the submitted code.

We're getting the pipeline fixed.  Board has approved the payment.

**Review by HyperPeek — COMMENTED** [review]
chainparam.cpp ln 165 / 333:
nStartTime and nTimeOut need updating

**Review by HyperPeek — COMMENTED** [review]
I like the cleanup in ismine.cpp for TX_NEW_ASSET / TX_TRANSFER_ASSET and TX_REISSUE_ASSET, it is much clearer this way.

**Review by HyperPeek — COMMENTED** [review]
Just some ocd -- standard.cpp ln 401 uses "scriptTyp", all other places use "scriptType"

**Review by HyperPeek — APPROVED** [review]
Very clean and nice code. I approve for testing in feature branch on testnet. May need changes to nStartTime as above.

File: src/chainparams.cpp [hunk 1]
```diff
@@ -161,6 +161,11 @@ class CMainParams : public CChainParams {
         consensus.vDeployments[Consensus::DEPLOYMENT_COINBASE_ASSETS].nTimeout = 1628877600; // UTC: Fri Aug 13 2021 18:00:00
         consensus.vDeployments[Consensus::DEPLOYMENT_COINBASE_ASSETS].nOverrideRuleChangeActivationThreshold = 1411; // Approx 70% of 2016
         consensus.vDeployments[Consensus::DEPLOYMENT_COINBASE_ASSETS].nOverrideMinerConfirmationWindow = 2016;
+        consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].bit = 11;
+        consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].nStartTime = 1617300000; // UTC: Thu Apr 01 2021 18:00:00
+        consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].nTimeout = 1648836000; // UTC: Fri Apr 01 2022 18:00:00
```
Comment (fdoving) [inline src/chainparams.cpp:166]: The dates need some modification.

File: src/test/assets/asset_p2sh_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        // Starting with mainnet addresses
+        // RRxxGQw8PWuGCDDcbF5x5euwiqn15zJUgD
+
+        // Create a transfer asset script to a non P2SH address
+        CAssetTransfer asset("NOTP2SHTEST", 1000);
+        CScript scriptPubKey = GetScriptForDestination(DecodeDestination("RRxxGQw8PWuGCDDcbF5x5euwiqn15zJUgD"));
+        asset.ConstructTransaction(scriptPubKey);
+
+        BOOST_CHECK_MESSAGE(!scriptPubKey.IsP2SHAssetScript(), "IsP2SHAssetScript should of failed to file a P2SH asset");
```
Comment (fdoving) [inline src/test/assets/asset_p2sh_tests.cpp:56]: s/of/have/ ?

File: src/consensus/tx_verify.cpp [hunk 3]
```diff
@@ -655,10 +655,16 @@ bool Consensus::CheckTxAssets(const CTransaction& tx, CValidationState& state, c
...
         int nType = 0;
+        int nScriptType = 0;
         bool fIsOwner = false;
-        if (txout.scriptPubKey.IsAssetScript(nType, fIsOwner))
+        if (txout.scriptPubKey.IsAssetScript(nType, nScriptType, fIsOwner))
             fIsAsset = true;
 
+        if (fIsAsset && nScriptType == TX_SCRIPTHASH) {
+            if (!AreP2SHAssetsAllowed())
+                return state.DoS(0, false, REJECT_INVALID, "bad-txns-p2sh-assets-not-active");
```
Comment (ben-abraham) [inline src/consensus/tx_verify.cpp:665]: Might be a good idea to move the DoS messages from choose to `BRConstants.h`, although it looks like most of the old ones are hard-coded already. Maybe a separate update.

File: src/script/script.cpp [hunk 4]
```diff
@@ -592,6 +624,27 @@ bool AmountFromReissueScript(const CScript& scriptPubKey, CAmount& nAmount)
     nAmount = asset.nAmount;
     return true;
 }
+
+int SearchForRVN(const CScript& script, const int startingValue) {
+
+    // Initialize the start value
+    int index = -1;
```
Comment (ben-abraham) [inline src/script/script.cpp:631]: Might be overkill, but would it be worth it to do a bounds check on `startingvalue` to make sure it doesn't exceed the size of `script` before the first check?

**Review by ben-abraham — APPROVED** [review]
Other than the already made comments everything here looks good! I haven't been able to test yet, but reading through all makes sense and no hidden gotchas that i could fine. Other than some constants and minor verbiage changes, looks good to me!

File: src/script/script.cpp (hunk 4 above)
Comment (HyperPeek) [reply src/script/script.cpp:631]: Agree -- will definitely not hurt.

File: src/chainparams.cpp [hunk 5]
```diff
@@ -161,7 +161,13 @@ class CMainParams : public CChainParams {
...
         consensus.vDeployments[Consensus::DEPLOYMENT_COINBASE_ASSETS].nOverrideRuleChangeActivationThreshold = 1411; // Approx 70% of 2016
         consensus.vDeployments[Consensus::DEPLOYMENT_COINBASE_ASSETS].nOverrideMinerConfirmationWindow = 2016;
-
+	consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].bit = 11;
+        consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].nStartTime = 1682956800; // UTC: Thu Mai 01 2023 18:00:00
+        consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].nTimeout = 1714579200; // UTC: Fri Mai 01 2024 18:00:00
+        consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].nOverrideRuleChangeActivationThreshold = 2016; // 100% required, hardly happens
+//        consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].nOverrideRuleChangeActivationThreshold = 1411; // Approx 70% of 2016
+        consensus.vDeployments[Consensus::DEPLOYMENT_P2SH_ASSETS].nOverrideMinerConfirmationWindow = 2016;
+
```
Comment (HyperPeek) [inline src/chainparams.cpp:170]: I changed settings of main net p2sh so it should never activate. Please double-check

**Pull request merged**